core/src/subgraph/context/instance.rs

```diff
@@ -155,7 +155,35 @@ where
         })
     }
 
-    pub(super) fn revert_data_sources(&mut self, reverted_block: BlockNumber) {
+    pub(super) fn revert_data_sources(
+        &mut self,
+        reverted_block: BlockNumber,
+    ) -> Vec<DataSource<C>> {
+        let mut removed = self.revert_quick(reverted_block);
+
+        // The following code handles resetting offchain datasources so in most
+        // cases this is enough processing.
+        // At some point we prolly need to improve the linear search but for now this
+        // should be fine. *IT'S FINE*
+        if self.causality_region_seq.0 == CausalityRegion::ONCHAIN {
+            return removed;
+        }
+
+        // TODO: replace this with drain_filter once it's stabilised.
+        let (drained, hosts): (Vec<_>, Vec<_>) =
+            self.hosts.iter().cloned().partition(
+                |host| matches!(host.done_at(), Some(done_at) if done_at >= reverted_block),
+            );
```

Comment:
I don't think we want to remove any hosts, but just set `done_at` to `None` on the necessary data sources.